### DISCUSSION TITLE
Trying to build a Korg Triton implementation - need help starting

### DISCUSSION THREAD

**Discussion by AJRubenstein** [christofmuc/KnobKraft-orm#139] [Q&A]
Hey all,

I've been trying to use the midi implementation that's listed on the Korg website, and I just can't get it to reply to Orm. I've watched Orm send the message through Midi-Ox and seen no reply from the Triton, but when I open up Triton Edit Pro (a SoundTower librarian/editor program) I see it send the exact same string of SysEx to the Triton and it reveals all of it's secrets. Same situation if I send through the request via SysEx in Midi-Ox. So I'm kinda mystified - what am I doing wrong here?

//   Built upon the Korg MS2000 file.
//
/ #   This works for program mode only, combination mode seems to be more complex to support

def name():
    return "Korg Triton Classic"

def createDeviceDetectMessage(channel):
    // This is a sysex generic device detect message
    return [0xf0, 0x7e, channel & 0x0f, 0x06, 0x01, 0xf7]

def deviceDetectWaitMilliseconds():
    return 200

def generalMessageDelay():
    # The 03R/W doesn't seem to like to get the messages too fast, so wait a bit between messages
    # The better implementation would be probably to do a handshake implementation, as it will likely reply
    # with a 0x23 DATA LOAD COMPLETED message on a program change message?
    return 400

def needsChannelSpecificDetection():
    return True

def channelIfValidDeviceResponse(message):

  //# 1-4-1 UNIVERSAL SYSTEM EXCLUSIVE MESSAGE ( NON REALTIME )
  //#      DEVICE INQUlRY REPLY ( Transmits when received a INQUIRY MESSAGE REQUEST )
  //#      [ F0,7E,0g,06,02,42,50,00,mm,00,nn,00,vv,00,F7 ]  3rd byte  g : Global Channel
  //#                                                        6th byte 42 : KORG ID
  //#                                                        7th byte 50 : TRITON series ID
  //#                                                        9th byte mm : TRITON      mm = 05
  //#                                                                      TRITON pro  mm = 0E
  //#                                                                      TRITON proX mm = 17
  //#                                                       11th byte nn : System No.      ( 01 -  )
  //#                                                       13th byte vv : System Version  ( 01 - 2 )

   // if (len(message) > 9
            and message[0] == 0xf0  # Sysex
            and message[1] == 0x7e  # Non-realtime
            and message[3] == 0x06  # Device request
            and message[4] == 0x02  # Device request reply
            and message[5] == 0x42  # KORG ID
            and message[6] == 0x50  # TRITON series ID
            and message[7] == 0x00
            and message[8] == 0x05  # TRITON
	    and message[9] == 0x00
            and message[12] == 0xF7):
      //  # Extract the current MIDI channel from index 2 of the message
        return message[2]
    return -1

This is all I've edited of the file so far. (sorry about all of the slashes, I couldn't work out how to turn off the markup!)

Thanks all,
AJ.

**Comment by Andy2No** [comment]
Is Orm detecting the synth?  At first glance, it looks like def channelIfValidDeviceResponse(message): might be wrong.

Try commenting out the last comparison line (put a # at the start):
```
# and message[12] == 0xF7):
```
then add the closing ): to the line above instead.

Failing that, do the same to some of the others, temporarily.  That could falsely identify a different synth as the Triton, but you can refine that later, once you see it working.

EDIT:  I missed the last part.  The way to paste in a block of code in a github comment is to use three back tick characters in a row, before and after it - it looks like a single quote mark, or an apostrophe, but it slopes the other way.  I always put those on a line of their own.

**Comment by christofmuc** [comment]
@AJRubenstein The MIDI log window and the console (the lower log view) of the Orm will be your friend. Carefully check the log window that there are no errors loading and starting your Python code, if any errors come up, they will be shown down there. And then print() in Python is redirected into the log window, so when you want to check your code (and make sure that it is actually run), use print() command liberally to see what is happening. It is pretty hard to attach a debugger to the embedded Python, so print debugging is probably what it is going to be ;-)

The Python code written BTW is in no way dependend on the Orm - you can also run and debug methods in a standalone python  editor (I always recommend PyCharm community edition). Have a look at the Matrix1000.py file in the Adaptations directory here on github and scroll to the very bottom of the file, you can see how I check that the script is run as a program and not from within the Orm, and how I use a hex string and convert it to binary data for testing my code: https://github.com/christofmuc/KnobKraft-orm/blob/master/adaptions/Matrix1000.py

Let us know if that helps, and feel free to post more output/MIDI logs in case you get stuck!

**Comment by AJRubenstein** [comment]
I am legitimately embarrassed at how long it's taken me to come back to this, but I think I have gotten somewhere with this adaption.

It's a hodgepodge of LLM garbage and my own tinkering, but at the moment I can at least use it to import banks and edit buffers, and I can send successfully (at least, one patch at a time) to the Triton. Attached to this post! I couldn't quite work out pull requests because it's past midnight over here.

I'm happy for it to be added to the pool - but I must ask, are there any other adaptions that have something similar to a 'combi' like the Triton? It's a grouping of Program patches to make one big patch... I feel I've only just broken the surface of how much I'll need to do to make this a properly functioning patch! haha..

[Korg_Triton.py](https://github.com/user-attachments/files/24150491/Korg_Triton.py)

**Reply by christofmuc** [reply]
Awesome! In the grander scheme of things, this is a mere blink of an eye! 

Happy you got it to run - what I have added in the meantime is a mechanism to have at learst some minimal tests running in that it can successfully open files and rename them. Do you have some, ideally publicly shareable, test files? One bank file would be completely enough, or mayby Korg even has the factory programs still somewhere? We can add them in as test data, which helps me maintain contributed adaptations should I ever need to change something for newer versions.

**Reply by AJRubenstein** [reply]
Hey, can do - what exactly are you asking me to share though? Like do I need to export it, or give you a sysex dump, or...?

**Reply by christofmuc** [reply]
Yeah, just a sysex file that we can hook up with some tests. You can see in the directory https://github.com/christofmuc/KnobKraft-orm/tree/master/adaptations/testData I collect a bit of real life data for each synth so we can make sure we don't break adaptations when we change the interface a bit.

**Reply by AJRubenstein** [reply]
No worries at all, I'll do a bank dump and a full sysex dump just so you have plenty of data to go on. This one is a bit of a complex demon of a device, I'll be completely frank! The Korg factory programs are still available but you need to load them in with their proprietary format (ie by 'floppy', despite the fact that I replaced the drive with a gotek years ago!) - but of course, you can also load the same data in via sysex but it needs to match the system version. 

Now, just to be clear, I'll do the sysex dumps via midi-ox just because I'm still not one hundred percent confident in my script, but in theory they should be the same, probably some time this evening. Cheers!

**Reply by AJRubenstein** [reply]
[triton.zip](https://github.com/user-attachments/files/24159551/triton.zip)
Hey, I've attached a single bank (what I'm currently successfully working with as a bank dump), a single patch, and a full sysex dump of the entire system. Hope that helps with future testing!

Do you have any suggestions on how to go about dumping 'combi' banks that have program pre-requisites? Is that a bit too grand for what Knobkraft ORM currently is capable of?

**Reply by christofmuc** [reply]
Thanks, I'll have a look at see if I can make the tests work! 

Regarding the Combi banks: I don't know the data format. There are many versions of multi/stack/layer combinations in the synths, and it is currently possible to archive those when carefully crafting the adaptation, but not necessarily possible to recombine e.g. two tones into a new program, or two layers into a new patch.

What does the Triton do in Combi mode?

**Reply by AJRubenstein** [reply]
I actually went pretty deep down the rabbit hole on this one last night. So here's what I got out of it.

First of all, the combi-banks are pretty simple at their heart, they're just a bunch of programs in one combi 'container' with set FX / Splits etc attached to them. I'm not really expecting Orm to handle that - but the bigger question is about whether the correct program patches are in place, and moreover, in the *right* place for the Combi banks to move across cleanly. Otherwise you'll end up with strings in your combi patch playing when it's expecting there to be drums or something like that. The idea is to be able to easily to splits or mix multiple sounds together to make one really nice one.

So I guess at a bare minimum I was hoping we could add some sort of information to a combi-bank patch that specifies what program patches need to be where. Here's some example combi patches - the 'timbres' are actually just program bank patches, the [A-033] specifies it expects to find 'Romance Piano' in Bank A patch 003, and finally the (CH17) I believe is what midi channel that program patch ('timbre') is using. 

Combi A-000: Romance Layers  
  Timbres:
    T1: Romance Piano    [A-033] (CH17)
    T2: AnalogStrings1&2 [A-099] (CH17)
    T3: AnalogStrings1&2 [A-099] (CH17)
    T4: AnalogStrings1&2 [A-099] (CH17)
    T5: AnalogStrings1&2 [A-099] (CH17)
    T6: AnalogStrings1&2 [A-099] (CH17)
    T7: <Off>
    T8: Percussion Kit   [B-116] (CH03)

Combi A-001: Lonely Moon     
  Timbres:
    T1: VenusianStories  [A-106] (CH17)
    T2: Dolphin Ride     [B-046] (CH17)
    T3: Trancer Lead     [A-060] (CH02)
    T4: Digi Syn Bass    [B-070] (CH17)
    T5: Dark R&B Bass    [B-006] (CH17)
    T6: <Off>
    T7: <Off>
    T8: <Off>

Some of these combi patches also use General Midi bank (Bank F) program patches as well, but those are always there.

According to the midi imp docs, the name of the patch is irrelevant, only where it is stored (program number / program bank number) and the value is stored as a single hex bit in the sysex message - ie :

```
+-----------+---------------------+-------------------------+-----------------------------+---------+ 
|    TIMBRE 1 PARAMETER                                                                             | 
+-----------+---------------------+-------------------------+-----------------------------+---------+ 
| 224       | PROGRAM NO.         |   00~~7F : 00~~127      |                             |   n,00  |  
+-----------+---------------------+-------------------------+-----------------------------+---------+ 
| 225       | PROGRAM BANK        |   00~~10 : Bank A~~g(d) |                             |   n,00  | 
+-----------+---------------------+-------------------------+-----------------------------+---------+ 

```

I've created a little python script for parsing the combi bank and program bank sysex to find what patches are needed and where, currently it requires you to have all of the banks dumped separately which is a bit of a pain. In theory I could also put together a script that rewrites the sysex values to change around the program numbers and program bank numbers to match the correct patch name... but I don't actually know how one could integrate that into Orm, as from what I gather the Python connection to the main program is mostly to allow certain functions to pull across to the main program. But if you have any insights on this... would love to hear them!

I've attached my python program and the separate banks if you ever have the time to look. Thanks for all of your insight so far.

[triton-sysexbanks-pythonscripts.zip](https://github.com/user-attachments/files/24178594/triton-sysexbanks-pythonscripts.zip)

**Reply by AJRubenstein** [reply]
@christofmuc I have gone a bit off the rails and I have pretty much mostly implemented all of the Orm features I could find!

[Korg_triton-adapt-withfingerprinting.py](https://github.com/user-attachments/files/24272223/Korg_triton-adapt-withfingerprinting.py)
This is a reasonably working version of the adaption that does work with said Combi files, using the layers system you've added to Orm, as well as fingerprinting.

At the moment it requires you to make a dump of a full set of program banks via sysex and convert them through a CSV table to get the program names (they're hex values in a table that point to Bank Number / Program Number for each 'layer') so it's a bit clunky, ideally I'll get the adaption working directly with pulling the individual programs for each combi... here's the program used at any rate, along with the sysex banks I've dumped:

[analyze_triton_multibank.py](https://github.com/user-attachments/files/24272291/analyze_triton_multibank.py)
[triton-sysexbanks.zip](https://github.com/user-attachments/files/24272301/triton-sysexbanks.zip)

but that leads me to a slightly deeper question.

I've built the fingerprinting hash so it doesn't acknowledge the program 'layers' within each combi quite deliberately as I was hoping there would be a way that we could write combi banks back to the synth while acknowledging the fingerprint for each program within the combi and programmatically assign the correct hex values to the combi before writing it back, so one could rearrange combi patches without having to ensure the program and bank match the original combi hex. Does this make sense at all? The fingerprinting not acknowledging the program bank/number for each layer means the hash is still correct no matter what the layer config is, and the idea would be that each layer would be matched up with each hash for it's respective program. This seems a little mad at the moment, but I would dearly love to implement it.

Either way, I'm having a lot of fun getting this to work!

**Reply by AJRubenstein** [reply]
@christofmuc Sorry about all of the little rants there, did you ever get a chance to test?

**Reply by christofmuc** [reply]
No worries, many things in parallel here as well. Yes, I think I had a go at it and it would work, I need to see I probably put then in a branch somewhere!

**Comment by christofmuc** [comment]
This the analysis result, there might be some work left for a full Korg Triton adaptation:

"• Findings first.

  1. The exposed bank list is wrong for a stock Triton Classic. D:/Development/github/KnobKraft-Orm-master/adaptations/
     Korg_Triton.py:139 only exposes INT-A through INT-D, but the manual’s program-bank footnotes show base program
     banks A-E, with F additionally used when available/MOSS is present (pages 9-10 in raw/TRITON.pdf). The request
     builder already accepts b = 0..5 in D:/Development/github/KnobKraft-Orm-master/adaptations/Korg_Triton.py:148, so
     the UI/API surface is narrower than the protocol. At minimum, bank E is missing.
  2. Single-program Triton dumps are supported by the spec and by the repo fixtures, but the adaptation does not model
     them as program dumps. The manual defines Func=1C with k=2 for “1 Program” and Func=4C as the corresponding
     response (page 6 and footnote *7 on page 9 in raw/TRITON.pdf). The shipped fixture D:/Development/github/KnobKraft-
     Orm-master/adaptations/testData/Korg_Triton/bank1-patch1-korgtriton-noisystabber.syx starts with 4C 00 20 00, which
     decodes as that one-program form. But the code only exposes bank requests in D:/Development/github/KnobKraft-Orm-
     master/adaptations/Korg_Triton.py:148 and treats every 4C as a bank dump in D:/Development/github/KnobKraft-Orm-
     master/adaptations/Korg_Triton.py:176. There is no createProgramDumpRequest, isSingleProgramDump, or
     numberFromDump, so an implemented Triton capability is currently left on the table.
  3. The adaptation is only partially implemented for Triton program data, not broadly “correctly implemented” yet. It
     hardcodes PCM program size 540 in D:/Development/github/KnobKraft-Orm-master/adaptations/Korg_Triton.py:24 and
     forces extracted programs to type 0x00 in D:/Development/github/KnobKraft-Orm-master/adaptations/
     Korg_Triton.py:253. That matches the supplied PCM bank fixtures, but it does not cover the manual’s wider program
     surface, especially the optional F/MOSS side described in raw/TRITON.pdf. The code comment on D:/Development/
     github/KnobKraft-Orm-master/adaptations/Korg_Triton.py:322 also admits the 0x50 all-data dump fixture is
     unimplemented.

  What did check out:

  - The current-program request F0 42 3g 50 10 00 F7 in D:/Development/github/KnobKraft-Orm-master/adaptations/
    Korg_Triton.py:73 matches the manual.
  - The Korg 7-bit unpacking logic is consistent with the sample bank dump. The one-bank fixture D:/Development/github/
    KnobKraft-Orm-master/adaptations/testData/Korg_Triton/bank1-korgtriton-midiox2.syx decodes to exactly 128 * 540
    bytes, and the first extracted name is Noisy Stabber.

  Validation:

  - I ran the shared adaptation suite from adaptations/ with pytest test_adaptations.py -q --adaptation Korg_Triton.py.
  - Result: 7 passed, 27 skipped. So nothing obvious is failing, but most of the useful surface is currently untested
    because the corresponding methods are missing.

  My read is: this is not wildly wrong, but it is incomplete. The safest next step is:

  1. fix bank descriptors to include at least INT-E,
  2. add real single-program support for the 4C k=2 form,
  3. decide whether bank F/MOSS is in scope or should be explicitly excluded."

**Comment by christofmuc** [comment]
@AJRubenstein I had Codex give it a go at implementing the missing bank dump imports. It was not shy at changing other parts of the code as well ;-)

If you want to give it a try or have a look, it is currently sitting in a pull request, the code is here: https://github.com/christofmuc/KnobKraft-orm/blob/3bb87039e4f0d7ef2792962ab1f1a4147e55d19e/adaptations/Korg_Triton.py
